src/Model/Filter/FilterInterface.php

```diff
@@ -0,0 +1,91 @@
+<?php
+namespace Search\Model\Filter;
+
+use Cake\Datasource\QueryInterface;
+
+/**
+ * Filter Interface
+ */
+interface FilterInterface
+{
+    /**
+     * Get the database field name(s) as an array.
+     *
+     * @return array
+     */
+    public function fields();
+
+    /**
+     * Get the field name from HTTP GET query string.
+     *
+     * @return string
+     */
+    public function name();
+
+    /**
+     * Check if the name is present in the arguments from HTTP GET.
+     *
+     * @return bool
+     */
+    public function present();
+
+    /**
+     * Check if empty value for name in query string should be filtered out.
+     *
+     * @return bool
+     */
+    public function filterEmpty();
+
+    /**
+     * Checks whether this finder should be skipped.
+     *
+     * @return bool
+     */
+    public function skip();
+
+    /**
+     * Get the value of the "name" from HTTP GET arguments.
+     *
+     * @return mixed
+     */
+    public function value();
+
+    /**
+     * Sets the args.
+     *
+     * @param array $args Value.
+     *
+     * @return void
+     */
+    public function setArgs(array $args);
+
+    /**
+     * Gets the args.
+     *
+     * @return array
+     */
+    public function getArgs();
+
+    /**
+     * Sets the query object.
+     *
+     * @param \Cake\Datasource\QueryInterface $value Value.
+     * @return void
+     */
+    public function setQuery(QueryInterface $value);
```

Comment:
> Query will always be implementation of QueryInterface.

Yes it will be but the issue is `ORM\Query` has additional methods which are not part of `QueryInterface`. That's the reason we can use the `QueryInterface` typehitn. Now we also want to allow this plugin to be usable with any data source which implements `RepositoryInterface` and not just `ORM\Table` as done through #203. The query objects for such repositories will only implement `Datasource\QueryInterface` (for e.g. muffin/webservice plugin). Hence the need for allowing `\Cake\Datasource\QueryInterface` or `\Cake\ORM\Query` instance.